src/v/archival/archiver_operations_impl.cc

```diff
@@ -299,6 +555,259 @@ class archiver_operations_impl : public archiver_operations_api {
         co_return co_await part->aborted_transactions(base, last);
     }
 
+    /// Generate and upload segment index
+    ///
+    /// The method consumes a segment as a stream. While the segment
+    /// index is generated it also computes segment stats. Then it compares
+    /// the stats with the metadata of the segment and returns error in case
+    //// of the mismatch. The last step is to upload the segment index.
+    ss::future<aggregated_upload_result> upload_segment_index(
+      std::reference_wrapper<retry_chain_node> workflow_rtc,
+      std::reference_wrapper<cloud_storage::segment_record_stats> stats,
+      reconciled_upload_candidate_ptr upload,
+      std::string_view index_path,
+      ss::input_stream<char> stream) noexcept {
+        try {
+            vlog(
+              _rtclog.debug, "creating remote segment index: {}", index_path);
+            auto base_kafka_offset = upload->metadata.base_kafka_offset();
+            constexpr auto index_sampling_step
+              = 64_KiB; // TODO: use proper constant
+
+            cloud_storage::offset_index ix{
+              upload->metadata.base_offset,
+              base_kafka_offset,
+              0,
+              index_sampling_step,
+              upload->metadata.base_timestamp};
+
+            auto builder = cloud_storage::make_remote_segment_index_builder(
+              upload->ntp,
+              std::move(stream),
+              ix,
+              upload->metadata.delta_offset,
+              index_sampling_step,
+              stats);
+
+            auto res = co_await builder->consume().finally(
+              [&builder] { return builder->close(); });
+
+            if (res.has_error()) {
+                vlog(
+                  _rtclog.error,
+                  "failed to create remote segment index: {}, error: {}",
+                  index_path,
+                  res.error());
+                co_return upload_result::failed;
+            }
+            // Compare stats to the expectation
+            // return upload_result.
+            const bool checks_disabled
+              = config::shard_local_cfg()
+                  .cloud_storage_disable_upload_consistency_checks.value();
+            if (!checks_disabled) {
+                // Compare against manifest
+                if (
+                  !upload->metadata.is_compacted
+                  && !segment_meta_matches_stats(
+                    upload->metadata, stats, _rtclog)) {
+                    co_return cloud_storage::upload_result::failed;
```

Comment:
should an error be logged here with cause of failure, or will it be logged in some other layer?

Reply:
I see it is logged in `segment_meta_matches_stats`